src/lib.rs

```diff
@@ -1,101 +1,405 @@
 #![allow(clippy::into_iter_on_ref)]
-mod args;
+mod config;
 mod context;
+mod macros;
 mod rule;
 mod rule_tester;
+mod rules;
 mod violation;
 
 use std::{
     borrow::Cow,
+    cmp::Ordering,
+    collections::HashMap,
+    fs,
+    ops::Deref,
+    path::{Path, PathBuf},
     process,
-    sync::atomic::{AtomicBool, Ordering},
+    sync::{Mutex, PoisonError},
 };
 
-pub use args::Args;
 use clap::Parser;
-use context::QueryMatchContext;
-use rule::{ResolvedRule, Rule, RuleBuilder, RuleListenerBuilder};
+pub use config::Config;
+use context::{PendingFix, QueryMatchContext};
+use rayon::prelude::*;
+use rule::{ResolvedRule, ResolvedRuleListener};
 pub use rule_tester::{RuleTestInvalid, RuleTester, RuleTests};
-use tree_sitter::{Node, Query};
-use violation::ViolationBuilder;
+use tree_sitter::Query;
+use violation::{ViolationBuilder, ViolationWithContext};
 
-use crate::context::Context;
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
 
-#[macro_export]
-macro_rules! regex {
-    ($re:expr $(,)?) => {{
-        static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
-        RE.get_or_init(|| regex::Regex::new($re).unwrap())
-    }};
+pub fn run_and_output(config: Config) {
+    let violations = run(config);
+    if violations.is_empty() {
+        process::exit(0);
+    }
+    for violation in violations {
+        violation.print();
+    }
+    process::exit(1);
 }
 
-const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
-const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
+const MAX_FIX_ITERATIONS: usize = 10;
+
+pub fn run(config: Config) -> Vec<ViolationWithContext> {
+    let resolved_rules = config.get_resolved_rules();
+    let aggregated_queries = AggregatedQueries::new(&resolved_rules, &config);
+    let tree_sitter_grep_args = get_tree_sitter_grep_args(&aggregated_queries);
+    let all_violations: Mutex<HashMap<PathBuf, Vec<ViolationWithContext>>> = Default::default();
```

Comment:
So nice to be able to just `default()` this.

Reply:
Without looking at what uses this, this smells like a place for a concurrent hashmap?

Reply:
Ya that sounds right the different threads would be "namespaced" so I assume would work nicely with that type of thing, added an issue